Custom training routines

dribnet opened the pull request:
These series of commits adds two new experiments which act as general purpose tools as discussed in #2, which this pull request is meant to replace. It adds a more generic version of train_classifier and train_vae, which are intended to be compatible with the celeba versions but with many additional options.
- Command line options for things that were hard coded like
  - --oldmodel to start training from a previously saved state
  - --classifier to use a different classifier filename
  - --model to use a different model filename
  - --batch-size to use different batch-sizes (generally smaller for memory reasons)
  - --z-dim to change network architecture allowing different sized latent space
  - --monitor-every and --checkpoint-every to change frequency of those events
- Split out `discriminative_term` in the cost function so it could be monitored separately
- Ability to scale relative cost of reconstruction, kl, and discriminative
- Classifier can train on a subset of labels given at runtime
- Can train against compatible fuel datasets other than CelebA

dribnet committed:
train_celeba_classifier: checkpointing and monitoring every epoch

Dropped checkpointing and monitoring in train_celeba_classifier.py
from every 5 epochs to every epoch - because the model
converges so quickly that even 5 seems like overkill.

dribnet committed:
train_celeba_classifier: added "allowed" label filter, cli opts

The training can now be done on a subset of the 40 available
CelebA labels. For example, to train only on "lipstick" and
"big lips", just provide the command line option with numbers
that match those columns:

  --allow 6,36

And all other labels will be zeroed out by a Transformer.

Also added the ability to specify other options on the command line:

  --classifier for output file, defaults to "celeba_classifier.zip"
  --batch-size for size of all mini-batches. defaults to 100

dribnet committed:
train_celeba_vae: added addition cli options

Added cli options:

  --classifer for input classifier, defaults to "celeba_classifier.zip"
  --model for output model, defaults to "celeba_vae_regularization.zip"
  --batch-size for all mini-batches, defaults to 100
  --z-dim to control latent dimensionality, defaults to 1000

dribnet committed:
train_celeba_vae: Separated discriminative_term for tracking

dribnet committed:
train_celeba_vae: added scaling factor options

Added three scaling factors with command line options to
change the relative weighting of the loss function:
reconstruction_factor, kl_factor, and discriminative_factor.

dribnet committed:
train_celeba_classifier: cli for checkpoint/monitor frequency

Added cli options --monitor-every and --checkpoint-every which
change how often monitoring and checkpointing occur during training.

dribnet committed:
train_celeba: allow custom dataset for classifier and vae

Added options to allow swapping out the celeba dataset with
any other fuel compatible dataset with 64x64 features. If
the dataset is grayscale instead of color, the color-convert
option can also be used to dynamically transform the
data into color as it comes in.

Note that for train_celeba_classifier the custom fuel dataset
must also have compatible targets, which is unlikely already
the case for common datasets but it certainly possible with
some customization. However, for train_celeba_vae any fuel
dataset with 64x64 features can be used - either with
regularize left off or by using an existing celeba classifier
or any other classifier trained on data close enough to
the target dataset.

dribnet committed:
train_celeba_vae: cli for checkpoint/monitor frequency

Added cli options --monitor-every and --checkpoint-every which
change how often monitoring and checkpointing occur during training.

dribnet committed:
experiments: rename generic trainers

dribnet committed:
recovered original train_celeb experiment scripts from 419687c2

dribnet committed:
train_vae: added --oldmodel option for starting from previous model

Added an option to start training from a prevoius model
checkpoint. Note that this also involved a change to MainLoop
to add the model, which sadly means that previously saved
checkpoints are not usable.

Also updated train_monitoring options to set
before_first_epoch=True as this is useful to verify
that the checkpoint was loaded successfully.

dribnet committed:
train_classifier: can restart training from saved, wider labels

Added --oldmodel to start training from a previously saved
model state.

Increased the width of the labels from 40 to 64 to support
different datasets. Will need to write a transformer to
deal with filling out from shorter label lengths.

dribnet committed:
utils.py: moved allowed option out of create_celeba_streams

Moved the allowed option used to filter labels from
create_celeba_streams to create_custom_streams.

vdumoulin commented on discgen/utils.py:
```diff
@@ -117,6 +124,79 @@ def create_svhn_streams(training_batch_size, monitoring_batch_size):
 
 
+class Colorize(AgnosticSourcewiseTransformer):
+    def __init__(self, data_stream, **kwargs):
```
Could you add a docstring explaining what this transformer does?

vdumoulin commented on discgen/utils.py:
```diff
@@ -117,6 +124,79 @@ def create_svhn_streams(training_batch_size, monitoring_batch_size):
+             **kwargs)
+
+    def transform_any_source(self, source, _):
+        iters, dim, height, width = source.shape
```
This will break if `source` is an example with shape `(dim, height, width)`.

dribnet replied:
does this happen if batch-size == 1?

vdumoulin replied:
Not as long as `source.ndim == 4`. The problem is if `source.ndim == 3`, it'll fail to unpack the 4 values you're asking.

dribnet replied:
If source.ndim == 3 for image data, would the fields then be:

``` python
iters, dim, length = source.shape
```

ie: a flat packing where length = width \* height? If so, agreed that would probably also be easy to support.

vdumoulin commented on discgen/utils.py:
```diff
@@ -117,6 +124,79 @@ def create_svhn_streams(training_batch_size, monitoring_batch_size):
+        cur_path_index = cur_path_index + 1
+    if not file_found:
+        print "data set {} not found, exiting".format(dataset_fname)
+        sys.exit(1)
```
I think the above block of code can be replaced by a call to `fuel.utils.find_in_data_path`.

vdumoulin commented on experiments/train_classifier.py:
```diff
@@ -0,0 +1,204 @@
+"""Trains the CelebA classifier."""
```
Maybe change this to something like `"""Trains a CelebA-like classifier."""`

vdumoulin commented on experiments/train_vae.py:
```diff
@@ -0,0 +1,466 @@
+"""Trains a VAE on the CelebA dataset."""
```
Maybe change this to `"""Trains a VAE on a CelebA-like dataset."""`?

vdumoulin commented on discgen/utils.py:
```diff
@@ -143,6 +223,18 @@ def create_cifar10_streams(training_batch_size, monitoring_batch_size):
     return create_streams(train_set, valid_set, test_set, training_batch_size,
                           monitoring_batch_size)
 
+class Scrubber(AgnosticSourcewiseTransformer):
```
Could you add a docstring explaining what this class does?

vdumoulin commented:
@dribnet I did a first review pass. Could you also `flake8` the files and fix issues that might have been introduced by your changes?

dribnet commented:
Thanks @vdumoulin for the constructive feedback, glad to hear you think overall this would be a welcome addition. I'll be reviewing and updating this PR over the next week.

dribnet committed:
light code cleanup: comments and find_in_data_path

Replaced block of custom code with fuel.utils.find_in_data_path
and added comments to several functions.

dribnet committed:
flake8 cleanup of new files

Formatting updates after running flake8 for better legibility.

dribnet committed:
train_vae: code cleanup using tensor.zeros_like

dribnet committed:
train_classifier.py: added --stretch option to zero pad labels

If incoming dataset doesn't provide enough labels for standard
training, they can be zero-padded with the --stretch option.

dribnet commented:
@vdumoulin - I've addressed most issues in the previous review including a general flake8 code cleanup.

dribnet committed:
train_vae: light formatting and arg help string
